lib/split-home-site-urls/asset-urls.php

```diff
@@ -0,0 +1,309 @@
+			$domain = $name . '.' . $tld;
+		} else {
+			$domain = $host;
+		}
+
+		// Array of Automattic domains
+		$domain_whitelist = array( 'wordpress.com', 'wp.com' );
+
+		// Return $uri if an Automattic domain, as it's already CDN'd
+		if ( in_array( $domain, $domain_whitelist, true ) ) {
+			return $uri;
+		}
+
+		// URI isn't served from a8c CDN, so serve from ours
+		return $this->staticize( $uri, current_filter() );
+	}
+
+	/**
+	 * Rewrite all upload URLs to point to the CDN
+	 * Ensures images not processed with Photon are at least served from the CDN
+	 *
+	 * @param string $url
+	 * @filter pre_option_upload_url_path
+	 * @return string
+	 */
+	public function filter_upload_url_path( $url ) {
+		$url = 'https://' . $this->static_host . '/wp-content/uploads';
+
+		if ( 2 === get_current_blog_id() ) {
+			$url .= '/sites/' . get_current_blog_id();
+		}
+
+		return esc_url_raw( $url );
+	}
+
+	/**
+	 * Rewrite URLs for concatenated assets
+	 *
+	 * @param string $url
+	 * @filter ngx_http_concat_site_url
+	 * @return string
+	 */
+	public function filter_concat_base_url( $url ) {
+		return 'https://' . $this->static_host;
+	}
+
+	/**
+	 ** UTILITY METHODS
+	 **/
+
+	/**
+	 * Restrict static-asset rewriting to "front-end" requests
+	 *
+	 * @return bool
+	 */
+	private function should_rewrite_non_upload_assets() {
+		// Allow dynamic exclusions
+		$override = apply_filters( 'wpcom_vip_asset_urls_skip_rewrites_for_request', null );
+		if ( is_bool( $override ) ) {
+			return $override;
+		}
+
+		// Admin should use its domain to avoid CORS, mixed content, and authentication issues, et al
+		if ( is_admin() ) {
+			return false;
+		}
+
+		// Previews use the site URL when this library is active
+		// Can't use is_preview() because this is called before the query is parsed
+		if ( isset( $_GET['preview'] ) ) {
+			return false;
+		}
+
+		// Skip on some Core front-end screens to avoid CORS and mixed content
+		if ( false !== stripos( $_SERVER['REQUEST_URI'], 'wp-login.php' ) ) {
+			return false;
+		}
+
+		if ( false !== stripos( $_SERVER['REQUEST_URI'], 'wp-signup.php' ) ) {
+			return false;
+		}
+
+		// Rewrite by default
+		return true;
+	}
+
+	/**
+	 * Rewrite host to static
+	 *
+	 * @param string $src
+	 * @param string $context Optional.
+	 * @return string
+	 */
+	private function staticize( $src, $context = '' ) {
+		// Abort if called too early
+		if ( ! $this->static_host ) {
+			return $src;
+		}
+
+		// Don't rewrite PHP URLs
+		$extension = pathinfo( $src, PATHINFO_EXTENSION );
+		if ( 0 === strpos( $extension, 'php' ) ) {
+			return $src;
+		}
+
+		// Attempt to rewrite if proper conditions are met.
+		$parsed_host = parse_url( $src, PHP_URL_HOST );
+		if ( $parsed_host && $parsed_host !== $this->static_host && $this->should_staticize( $parsed_host ) && apply_filters( 'wpcom_vip_asset_urls_staticize', true, $parsed_host, $src, $context ) ) {
+			$src = str_replace( $parsed_host, $this->static_host, $src );
```

Comment:
We should put a limit of `1` on the `str_replace()`, so that we don't affect other parts of the url, like GET params - we just want to change the hostname.